experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+import AbstractAlgebra.Ring, Oscar.AlgHom, Oscar.compose, AbstractAlgebra.Generic.Frac
+import Base: ∘
+import Oscar: base_ring
+import AbstractAlgebra: FPModule, FPModuleElem
+
+include( "./Multiindices.jl" )
+using Oscar.Multiindices
+
+include( "./Misc.jl" )
+using Oscar.Misc
+
+export AffineScheme, PrincipalSubScheme,Spec, SpecPrincipalOpen, affine_space
+export AffSchMorphism, base_ring,ambient_ring,defining_ideal, pullback, pullback_from_parent, pullback_from_root, inclusion_in_parent, inclusion_in_root, set_name!
+export localize, get_unit
+
+export AffSchMorphism
+export domain, codomain, pullback
+
+export Glueing
+export first_patch, second_patch, overlap, first_inclusion, second_inclusion, glueing_isomorphism
+
+export Covering
+export get_patches, get_glueings, add_patch
+
+export Refinement
+
+export CoveredScheme
+export get_coverings
+
+export AbstractCoherentSheaf
+
+export IdealSheaf
+export parent, covering, ideals
+
+export CoherentSheaf
+export parent, covering, modules
+
+export LineBundle
+export parent, covering, transitions, OO
+
+abstract type Scheme{ S <: Ring }end
+abstract type AffineScheme{S, T <: MPolyRing, U <: MPolyElem} <: Scheme{S} end
+abstract type SchemeMorphism end
+
+abstract type SchematicPoint end
+abstract type Sheaf end
+abstract type AbstractCoherentSheaf <: Sheaf end
+
+####################################################################################
+# The classical affine scheme, explicitly given as the quotient 
+# ring of a polynomial algebra.
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
```

Comment:
You can restrict the acceptable type parameters here
```suggestion
mutable struct Spec{S<:Ring,T<:MPolyRing,U<:MPolyElem} <: AffineScheme{S,T,U}
```

Reply:
Ok... But have they not already been restricted in the declaration of AffineScheme? Why do that twice?

Reply:
Sorry, I missed the definition of `AffineScheme` somehow.